Content.Server/Explosion/EntitySystems/ExplosionSystem.GridMap.cs

```diff
@@ -0,0 +1,386 @@
+                {
+                    var oppositeDirection = (NeighborFlag) (1 << ((i + 4) % 8));
+                    edges[neighbourIndex] = edges.GetValueOrDefault(neighbourIndex) | oppositeDirection;
+                }
+            }
+
+            return;
+        }
+
+        // the tile is not empty space, but was previously. So update directly adjacent neighbours, which may no longer
+        // be edge tiles.
+        for (var i = 0; i < NeighbourVectors.Length; i++)
+        {
+            var neighbourIndex = tileRef.GridIndices + NeighbourVectors[i];
+
+            if (edges.TryGetValue(neighbourIndex, out var neighborSpaceDir))
+            {
+                var oppositeDirection = (NeighborFlag) (1 << ((i + 4) % 8));
+                neighborSpaceDir &= ~oppositeDirection;
+                if (neighborSpaceDir == NeighborFlag.Invalid)
+                {
+                    // no longer an edge tile
+                    edges.Remove(neighbourIndex);
+                    continue;
+                }
+
+                edges[neighbourIndex] = neighborSpaceDir;
+            }
+        }
+
+        // finally check if the new tile is itself an edge tile
+        if (IsEdge(grid, tileRef.GridIndices, out var spaceDir))
+            edges.Add(tileRef.GridIndices, spaceDir);
+    }
+
+    /// <summary>
+    ///     Check whether a tile is on the edge of a grid (i.e., whether it borders space).
+    /// </summary>
+    /// <remarks>
+    ///     Optionally ignore a specific Vector2i. Used by <see cref="OnTileChanged"/> when we already know that a
+    ///     given tile is not space. This avoids unnecessary TryGetTileRef calls.
+    /// </remarks>
+    private bool IsEdge(IMapGrid grid, Vector2i index, out NeighborFlag spaceDirections)
+    {
+        spaceDirections = NeighborFlag.Invalid;
+        for (var i = 0; i < NeighbourVectors.Length; i++)
+        {
+            if (!grid.TryGetTileRef(index + NeighbourVectors[i], out var neighborTile) || neighborTile.Tile.IsEmpty)
+                spaceDirections |= (NeighborFlag) (1 << i);
+        }
+
+        return spaceDirections != NeighborFlag.Invalid;
+    }
+
+    // yeah this is now the third direction flag enum, and the 5th (afaik) direction enum overall.....
+    /// <summary>
+    ///     Directional bitflags used to denote the neighbouring tiles of some tile on a grid.. Differ from atmos and
+    ///     normal directional flags as NorthEast != North | East
+    /// </summary>
+    [Flags]
+    public enum NeighborFlag : byte
+    {
+        Invalid = 0,
+        North = 1 << 0,
+        NorthEast = 1 << 1,
+        East = 1 << 2,
+        SouthEast = 1 << 3,
+        South = 1 << 4,
+        SouthWest = 1 << 5,
+        West = 1 << 6,
+        NorthWest = 1 << 7,
+
+        Cardinal = North | East | South | West,
+        Diagonal = NorthEast | SouthEast | SouthWest | NorthWest,
+        Any = Cardinal | Diagonal
+    }
```

Comment:
is there REALLY no way to use one of the other direction flags? why cant `NorthEast = North | East`

Reply:
Because these flags need to be able to address each of the 7 neighbouring tiles separately. I.e., you need to be able to refer to the north-eastern tile separately from the eastern or northern tiles. Unless I've missed a direction enum somewhere, I don't think any of the existing enum flags let you do that.

Reply:
thats ass but understandable